jerry-core/ecma/base/ecma-helpers-string.c

```diff
@@ -90,6 +84,69 @@ ecma_init_ecma_string_from_magic_string_ex_id (ecma_string_t *string_p, /**< des
 } /* ecma_init_ecma_string_from_magic_string_ex_id */
 
 /**
+ * Convert a string to an unsigned 32 bit value if possible
+ *
+ * @return true if the conversion is successful
+ *         false otherwise
+ */
+static bool
+ecma_string_to_array_index (const lit_utf8_byte_t *string_p, /**< utf-8 string */
+                            lit_utf8_size_t string_size, /**< string size */
+                            uint32_t *result_p) /**< [out] converted value */
+{
+  JERRY_ASSERT (string_size > 0 && *string_p >= LIT_CHAR_0 && *string_p <= LIT_CHAR_9);
+
+  if (*string_p == LIT_CHAR_0)
+  {
+    *result_p = 0;
+    return (string_size == 1);
+  }
+
+  if (string_size > ECMA_MAX_CHARS_IN_STRINGIFIED_UINT32)
+  {
+    return false;
+  }
+
+  uint32_t index = 0;
+  const lit_utf8_byte_t *string_end_p = string_p + string_size;
+
+  if (string_size == ECMA_MAX_CHARS_IN_STRINGIFIED_UINT32)
+  {
+    string_end_p--;
+  }
+
+  do
+  {
+    if (*string_p > LIT_CHAR_9 || *string_p < LIT_CHAR_0)
+    {
+      return false;
+    }
+
+    index = (index * 10) + (uint32_t) (*string_p++ - LIT_CHAR_0);
+  }
+  while (string_p < string_end_p);
+
+  if (string_size < ECMA_MAX_CHARS_IN_STRINGIFIED_UINT32)
```

Comment:
This is a tricky code. The index computation cannot overflow for numbers whose length <= 9 characters. But for exactly 10 character long numbers we need to check overflow. This is why we have many checks, and we need them.